src/geo_traits_impl.rs

```diff
@@ -0,0 +1,717 @@
+                } else {
+                    panic!()
+                }
+            }
+            _ => panic!("invalid dimension index"),
+        }
+    }
+
+    fn x(&self) -> Self::T {
+        self.x
+    }
+
+    fn y(&self) -> Self::T {
+        self.y
+    }
+}
+
+impl CoordTrait for &PointZ {
+    type T = f64;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        if self.m <= NO_DATA {
+            geo_traits::Dimensions::Xyz
+        } else {
+            geo_traits::Dimensions::Xyzm
+        }
+    }
+
+    fn nth_or_panic(&self, n: usize) -> Self::T {
+        match n {
+            0 => self.x(),
+            1 => self.y(),
+            2 => self.z,
+            3 => {
+                if self.m > NO_DATA {
+                    self.m
+                } else {
+                    panic!()
+                }
+            }
+            _ => panic!("invalid dimension index"),
+        }
+    }
+
+    fn x(&self) -> Self::T {
+        self.x
+    }
+
+    fn y(&self) -> Self::T {
+        self.y
+    }
+}
+
+impl PointTrait for PointZ {
+    type T = f64;
+    type CoordType<'a> = &'a PointZ where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        if self.m <= NO_DATA {
+            geo_traits::Dimensions::Xyz
+        } else {
+            geo_traits::Dimensions::Xyzm
+        }
+    }
+
+    fn coord(&self) -> Option<Self::CoordType<'_>> {
+        Some(self)
+    }
+}
+
+impl PointTrait for &PointZ {
+    type T = f64;
+    type CoordType<'a> = &'a PointZ where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        if self.m <= NO_DATA {
+            geo_traits::Dimensions::Xyz
+        } else {
+            geo_traits::Dimensions::Xyzm
+        }
+    }
+
+    fn coord(&self) -> Option<Self::CoordType<'_>> {
+        Some(self)
+    }
+}
+
+pub struct LineString<'a>(&'a [Point]);
+
+impl<'a> LineStringTrait for LineString<'a> {
+    type T = f64;
+    type CoordType<'b> = &'b Point where Self: 'b;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xy
+    }
+
+    fn num_coords(&self) -> usize {
+        self.0.len()
+    }
+
+    unsafe fn coord_unchecked(&self, i: usize) -> Self::CoordType<'_> {
+        &self.0[i]
+    }
+}
+
+pub struct LineStringM<'a>(&'a [PointM]);
+
+impl<'a> LineStringTrait for LineStringM<'a> {
+    type T = f64;
+    type CoordType<'b> = &'b PointM where Self: 'b;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xym
+    }
+
+    fn num_coords(&self) -> usize {
+        self.0.len()
+    }
+
+    unsafe fn coord_unchecked(&self, i: usize) -> Self::CoordType<'_> {
+        &self.0[i]
+    }
+}
+
+pub struct LineStringZ<'a>(&'a [PointZ]);
+
+impl<'a> LineStringTrait for LineStringZ<'a> {
+    type T = f64;
+    type CoordType<'b> = &'b PointZ where Self: 'b;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        // TODO: actually check whether M value exists
+        geo_traits::Dimensions::Xyz
```

Comment:
Should we act on these TODO ?

Reply:
In general it can be hard to know the correct dimension because shapefile overloads a single type to hold both Z and ZM data. I updated this to check the first coordinate of `Z` types, and use that as the dimension of the geometry.